gno.land/cmd/genesis/balances_add.go

```diff
@@ -0,0 +1,346 @@
+	}
+
+	return nil
+}
+
+// getBalancesFromEntries extracts the balance entries
+// from the array of balance
+func getBalancesFromEntries(entries []string) (accountBalances, error) {
+	balances := make(accountBalances)
+
+	for _, entry := range entries {
+		accountBalance, err := getBalanceFromEntry(entry)
+		if err != nil {
+			return nil, fmt.Errorf("unable to extract balance data, %w", err)
+		}
+
+		balances[accountBalance.address] = accountBalance.amount
+	}
+
+	return balances, nil
+}
+
+// getBalancesFromSheet extracts the balance sheet from the passed in
+// balance sheet file, that has the format of <address>=<amount>ugnot
+func getBalancesFromSheet(sheet io.Reader) (accountBalances, error) {
+	// Parse the balances
+	balances := make(accountBalances)
+	scanner := bufio.NewScanner(sheet)
+
+	for scanner.Scan() {
+		entry := scanner.Text()
+
+		// Remove comments
+		entry = strings.TrimSpace(entry)
+		entry = strings.Split(entry, "#")[0]
+		entry = strings.TrimSpace(entry)
+
+		// Skip empty lines
+		if entry == "" {
+			continue
+		}
+
+		accountBalance, err := getBalanceFromEntry(entry)
+		if err != nil {
+			return nil, fmt.Errorf("unable to extract balance data, %w", err)
+		}
+
+		balances[accountBalance.address] = accountBalance.amount
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("error encountered while scanning, %w", err)
+	}
+
+	return balances, nil
+}
+
+// getBalancesFromTransactions constructs a balance map based on MsgSend messages.
+// This way of determining the final balance sheet is not valid, since it doesn't take into
+// account different message types (ex. MsgCall) that can initialize accounts with some balance values.
+// The right way to do this sort of initialization is to spin up an in-memory node
+// and execute the entire transaction history to determine touched accounts and final balances,
+// and construct a balance sheet based off of this information
+func getBalancesFromTransactions(ctx context.Context, reader io.Reader) (accountBalances, error) {
+	balances := make(accountBalances)
+
+	scanner := bufio.NewScanner(reader)
+
+	for scanner.Scan() {
+		select {
+		case <-ctx.Done():
+			return nil, errBalanceParsingAborted
+		default:
+			// Parse the amino JSON
+			var tx std.Tx
+
+			if err := amino.UnmarshalJSON(scanner.Bytes(), &tx); err != nil {
+				return nil, fmt.Errorf(
+					"unable to unmarshal amino JSON, %w",
+					err,
+				)
+			}
+
+			feeAmount, err := getAmountFromEntry(tx.Fee.GasFee.String())
+			if err != nil {
+				return nil, fmt.Errorf(
+					"invalid gas fee amount, %s",
+					tx.Fee.GasFee.String(),
+				)
+			}
+
+			for _, msg := range tx.Msgs {
+				if msg.Type() != "send" {
+					continue
+				}
+
+				msgSend := msg.(bank.MsgSend)
+
+				sendAmount, err := getAmountFromEntry(msgSend.Amount.String())
+				if err != nil {
+					return nil, fmt.Errorf(
+						"%s, %s",
+						"invalid send amount",
+						msgSend.Amount.String(),
+					)
+				}
```

Comment:
@moul 

Just a note on this line of the parsing: this will fail if it encounters any message send that is not `<amount>ugnot`.

For some weird reason, the transactions on our `tx-exports` repo have transactions that have sent ex `10uatom` in the past (test3). These exports are not parsable in this state.

Do you think we should ignore messages that are "malformed" like this, or do you think we should keep the hard stop in processing?

Reply:
Yes, ignite with a warning

Reply:
Updated the parsing to just ignore invalid transactions / messages:

[5f1a43b](https://github.com/gnolang/gno/pull/1276/commits/5f1a43b23ad003e1fad1e1439dd1aa29261fff49)